In [1]:
import numpy as np
import pandas as pd

# Import Data

In [2]:
df = pd.read_csv(r"C:\Users\kunwar gupta\Downloads\Direct-to-Consumer E-Commerce Funnel Dataset\d2c_marketing_funnel_data.csv")

In [3]:
df.head()

,user_id,session_id,date,month,channel,campaign_type,device,user_type,region,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,discount_applied,order_value,revenue
0,221958,1,2025-08-16,08-25,Organic,New Launch,Mobile,New,Metro,Yes,No,No,No,No,No,499.00,0.000
1,771155,2,2025-12-16,12-25,Organic,Influencer,Mobile,New,Non-Metro,Yes,Yes,Yes,No,No,No,499.00,0.000
2,231932,3,2025-07-17,07-25,Organic,Influencer,Mobile,New,Non-Metro,Yes,Yes,No,No,No,No,499.00,0.000
3,465838,4,2025-07-04,07-25,Paid Ads,Discount,Mobile,Returning,Metro,Yes,Yes,Yes,Yes,Yes,Yes,2000.95,1800.855
4,359178,5,2025-08-10,08-25,Paid Ads,Influencer,Mobile,Returning,Non-Metro,Yes,No,No,No,No,No,499.00,0.000


# Data Exploration

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120000 entries, 0 to 119999
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   user_id             120000 non-null  int64  
 1   session_id          120000 non-null  int64  
 2   date                120000 non-null  object 
 3   month               120000 non-null  object 
 4   channel             120000 non-null  object 
 5   campaign_type       120000 non-null  object 
 6   device              120000 non-null  object 
 7   user_type           120000 non-null  object 
 8   region              120000 non-null  object 
 9   visited_website     120000 non-null  object 
 10  viewed_product      120000 non-null  object 
 11  added_to_cart       120000 non-null  object 
 12  checkout_started    120000 non-null  object 
 13  purchase_completed  120000 non-null  object 
 14  discount_applied    120000 non-null  object 
 15  order_value         120000 non-nul

In [5]:
df.isnull().sum()

user_id               0
session_id            0
date                  0
month                 0
channel               0
campaign_type         0
device                0
user_type             0
region                0
visited_website       0
viewed_product        0
added_to_cart         0
checkout_started      0
purchase_completed    0
discount_applied      0
order_value           0
revenue               0
dtype: int64

# Data Cleaning & Conversion to Binary

In [6]:
df['visited_website'].unique()

array(['Yes'], dtype=object)

In [7]:
df['visited_website'] = df['visited_website'].map({"Yes": 1})

In [8]:
df['viewed_product'].unique()

array(['No', 'Yes'], dtype=object)

In [9]:
df['viewed_product'] = df['viewed_product'].map({"Yes": 1, "No":0})

In [10]:
df['added_to_cart'].unique()

array(['No', 'Yes'], dtype=object)

In [11]:
df['added_to_cart'] = df['added_to_cart'].map({"Yes": 1, "No": 0})

In [12]:
df["checkout_started"].unique()

array(['No', 'Yes'], dtype=object)

In [13]:
df["checkout_started"] = df["checkout_started"].map({"Yes": 1, "No": 0})

In [14]:
df["discount_applied"].unique()

array(['No', 'Yes'], dtype=object)

In [15]:
df["discount_applied"] = df["discount_applied"].map({"Yes": 1, "No": 0})

In [16]:
df["purchase_completed"].unique()

array(['No', 'Yes'], dtype=object)

In [17]:
df["purchase_completed"] = df["purchase_completed"].map({"Yes": 1, "No": 0})

In [18]:
df.head()

,user_id,session_id,date,month,channel,campaign_type,device,user_type,region,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,discount_applied,order_value,revenue
0,221958,1,2025-08-16,08-25,Organic,New Launch,Mobile,New,Metro,1,0,0,0,0,0,499.00,0.000
1,771155,2,2025-12-16,12-25,Organic,Influencer,Mobile,New,Non-Metro,1,1,1,0,0,0,499.00,0.000
2,231932,3,2025-07-17,07-25,Organic,Influencer,Mobile,New,Non-Metro,1,1,0,0,0,0,499.00,0.000
3,465838,4,2025-07-04,07-25,Paid Ads,Discount,Mobile,Returning,Metro,1,1,1,1,1,1,2000.95,1800.855
4,359178,5,2025-08-10,08-25,Paid Ads,Influencer,Mobile,Returning,Non-Metro,1,0,0,0,0,0,499.00,0.000


In [19]:
df['date'] = pd.to_datetime(df['date'])

In [20]:
df.dtypes

user_id                        int64
session_id                     int64
date                  datetime64[ns]
month                         object
channel                       object
campaign_type                 object
device                        object
user_type                     object
region                        object
visited_website                int64
viewed_product                 int64
added_to_cart                  int64
checkout_started               int64
purchase_completed             int64
discount_applied               int64
order_value                  float64
revenue                      float64
dtype: object

In [21]:
df.isnull().sum()

user_id               0
session_id            0
date                  0
month                 0
channel               0
campaign_type         0
device                0
user_type             0
region                0
visited_website       0
viewed_product        0
added_to_cart         0
checkout_started      0
purchase_completed    0
discount_applied      0
order_value           0
revenue               0
dtype: int64

# Logic Verification

In [22]:
# Can someone checkout without adding cart?

((df['checkout_started'] == 1) & (df['added_to_cart'] == 0)).sum()

np.int64(0)

In [23]:
# If purchased, revenue > 0?

((df['purchase_completed'] == 1) & (df['revenue'] == 0)).sum()

np.int64(0)

In [24]:
# Can someone purchase without checkout?

((df['checkout_started'] == 0) & (df['purchase_completed'] == 1)).sum()

np.int64(0)

## Saving Clean Data

In [25]:
df.to_csv('cleaned_data.csv', index=False)

# Phase 2: Funnel Analysis

In [26]:
# 1. Count visitors at each stage

# Stage1: count of website visitors
visited = df['visited_website'].sum()

# Stage 2: count of product viewed
viewed = df['viewed_product'].sum()

# Stage 3: count of product added to cart
cart = df['added_to_cart'].sum()

# Stage 4: count of checkout
checkout = df['checkout_started'].sum()

# Stage 5: count of purchased
purchased = df['purchase_completed'].sum()

In [27]:
# Calculate drop rates

drop_visited_to_viewed = ((visited - viewed) / visited) * 100
drop_viewed_to_cart = ((viewed - cart) / viewed) * 100
drop_cart_to_checkout = ((cart - checkout) / cart) * 100
drop_checkout_to_purchase = ((checkout - purchased) / checkout) * 100


In [28]:
# 3.  Calculate retention %

viewed_retention = (viewed / visited) * 100
cart_retention = (cart / viewed) * 100
checkout_retention = (checkout / cart) * 100
purchase_retention = (purchased / checkout) * 100


In [59]:
round(viewed_retention, 2), round(cart_retention, 2), round(checkout_retention, 2), round(purchase_retention, 2)

(np.float64(64.89), np.float64(34.87), np.float64(59.78), np.float64(50.39))

In [29]:
# 4. Create results table

result = {
    "Stage": ["Visited", "Viewed", "Cart", "Checkout", "Purchased"],
    "Count": [visited, viewed, cart, checkout, purchased],
    "Drop": [np.nan, drop_visited_to_viewed, drop_viewed_to_cart, drop_cart_to_checkout, drop_checkout_to_purchase],
    "Retention": [np.nan, viewed_retention, cart_retention, checkout_retention, purchase_retention]
}

In [30]:
funnel_df = pd.DataFrame(result)
funnel_df

,Stage,Count,Drop,Retention
0,Visited,120000,NaN,NaN
1,Viewed,77870,35.108333,64.891667
2,Cart,27156,65.126493,34.873507
3,Checkout,16234,40.219473,59.780527
4,Purchased,8181,49.605766,50.394234


In [61]:

stages = {
    "visited": "visited_website",
    "viewed": "viewed_product",
    "cart": "added_to_cart",
    "checkout": "checkout_started",
    "purchased": "purchase_completed"
}

def stage_retention(data, group_cols):
    result = data.groupby(group_cols, as_index=False).agg({
        stages["visited"]: "sum",
        stages["viewed"]: "sum",
        stages["cart"]: "sum",
        stages["checkout"]: "sum",
        stages["purchased"]: "sum"
    })

    result = result.rename(columns={
        stages["visited"]: "visited",
        stages["viewed"]: "viewed",
        stages["cart"]: "cart",
        stages["checkout"]: "checkout",
        stages["purchased"]: "purchased"
    })

    result["visit_to_view_retention"] = (
        result["viewed"] / result["visited"] * 100
    )

    result["view_to_cart_retention"] = (
        result["cart"] / result["viewed"] * 100
    )

    result["cart_to_checkout_retention"] = (
        result["checkout"] / result["cart"] * 100
    )

    result["checkout_to_purchase_retention"] = (
        result["purchased"] / result["checkout"] * 100
    )

    result["conversion_rate"] = (
        result["purchased"] / result["visited"] * 100
    )

    percentage_cols = [
        "visit_to_view_retention",
        "view_to_cart_retention",
        "cart_to_checkout_retention",
        "checkout_to_purchase_retention",
        "conversion_rate"
    ]

    result[percentage_cols] = result[percentage_cols].round(2)

    return result

In [62]:
channel_retention = stage_retention(df, ["channel"])
device_retention = stage_retention(df, ["device"])
user_type_retention = stage_retention(df, ["user_type"])
campaign_retention = stage_retention(df, ["campaign_type"])
region_retention = stage_retention(df, ["region"])

In [63]:
print(channel_retention)
print(device_retention)
print(user_type_retention)
print(campaign_retention)
print(region_retention)

    channel  visited  viewed   cart  checkout  purchased  \
0     Email    12092    7892   2838      1722        884   
1   Organic    35946   23327   8158      4909       2448   
2  Paid Ads    53891   35011  12058      7161       3619   
3    Social    18071   11640   4102      2442       1230   

   visit_to_view_retention  view_to_cart_retention  \
0                    65.27                   35.96   
1                    64.89                   34.97   
2                    64.97                   34.44   
3                    64.41                   35.24   

   cart_to_checkout_retention  checkout_to_purchase_retention  conversion_rate  
0                       60.68                           51.34             7.31  
1                       60.17                           49.87             6.81  
2                       59.39                           50.54             6.72  
3                       59.53                           50.37             6.81  
    device  visited  vi

In [64]:
channel_retention.to_csv("retention_by_channel.csv", index=False)
device_retention.to_csv("retention_by_device.csv", index=False)
user_type_retention.to_csv("retention_by_user_type.csv", index=False)
campaign_retention.to_csv("retention_by_campaign.csv", index=False)
region_retention.to_csv("retention_by_region.csv", index=False)

In [31]:
funnel_df.to_csv("funnel_analysis.csv", index = False)

In [32]:
funnel_by_channel = df.groupby('channel').agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
})

In [33]:
funnel_by_channel['conversion_rate'] = (funnel_by_channel['purchase_completed']/funnel_by_channel['visited_website']) * 100

funnel_by_channel

,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
channel,,,,,,
Email,12092,7892,2838,1722,884,7.310619
Organic,35946,23327,8158,4909,2448,6.810215
Paid Ads,53891,35011,12058,7161,3619,6.715407
Social,18071,11640,4102,2442,1230,6.806486


In [34]:
funnel_by_device = df.groupby('device').agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
})


In [35]:
funnel_by_device['conversion_rate'] = (funnel_by_device['purchase_completed']/funnel_by_device['visited_website']) * 100
funnel_by_device

,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
device,,,,,,
Desktop,35994,23373,8140,4893,2488,6.912263
Mobile,84006,54497,19016,11341,5693,6.776897


In [36]:
funnel_by_user_type = df.groupby('user_type').agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
})


In [37]:
funnel_by_user_type['conversion_rate'] = (funnel_by_user_type['purchase_completed']/funnel_by_user_type['visited_website']) * 100
funnel_by_user_type

,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
user_type,,,,,,
New,77969,50616,17730,10658,5398,6.923264
Returning,42031,27254,9426,5576,2783,6.621303


In [38]:
funnel_by_campaign = df.groupby('campaign_type').agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
})

In [39]:
funnel_by_campaign['conversion_rate'] = (funnel_by_campaign['purchase_completed']/funnel_by_campaign['visited_website']) * 100
funnel_by_campaign

,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
campaign_type,,,,,,
Discount,60121,38950,13541,8087,4107,6.831224
Influencer,23815,15423,5465,3240,1626,6.827630
New Launch,36064,23497,8150,4907,2448,6.787933


In [40]:
funnel_by_region = df.groupby('region').agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
})

In [41]:
funnel_by_region['conversion_rate'] = (funnel_by_region['purchase_completed']/funnel_by_region['visited_website']) * 100
funnel_by_region

,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
region,,,,,,
Metro,72014,46651,16181,9643,4909,6.816730
Non-Metro,47986,31219,10975,6591,3272,6.818655


# Phase 3: Cross-segment analysis

In [42]:
funnel_by_channel_device = df.groupby(['channel', 'device']).agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
}).reset_index()

In [43]:
funnel_by_channel_device['conversion_rate'] = (funnel_by_channel_device['purchase_completed']/funnel_by_channel_device['visited_website']) * 100
funnel_by_channel_device

,channel,device,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
0,Email,Desktop,3565,2309,870,550,290,8.134642
1,Email,Mobile,8527,5583,1968,1172,594,6.966108
2,Organic,Desktop,10825,7011,2432,1443,740,6.836028
3,Organic,Mobile,25121,16316,5726,3466,1708,6.799092
4,Paid Ads,Desktop,16209,10588,3630,2158,1089,6.718490
5,Paid Ads,Mobile,37682,24423,8428,5003,2530,6.714081
6,Social,Desktop,5395,3465,1208,742,369,6.839666
7,Social,Mobile,12676,8175,2894,1700,861,6.792364


In [44]:
monthly_trend = df.groupby(['channel', 'user_type']).agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
}).reset_index()

In [45]:
monthly_trend['conversion_rate'] = (monthly_trend['purchase_completed']/monthly_trend['visited_website']) * 100
monthly_trend

,channel,user_type,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
0,Email,New,7917,5207,1896,1158,602,7.603890
1,Email,Returning,4175,2685,942,564,282,6.754491
2,Organic,New,23440,15255,5352,3203,1592,6.791809
3,Organic,Returning,12506,8072,2806,1706,856,6.844715
4,Paid Ads,New,34887,22595,7828,4688,2400,6.879353
5,Paid Ads,Returning,19004,12416,4230,2473,1219,6.414439
6,Social,New,11725,7559,2654,1609,804,6.857143
7,Social,Returning,6346,4081,1448,833,426,6.712890


In [46]:
monthly_trend = df.groupby(['device', 'campaign_type']).agg({
    'visited_website': 'sum',
    'viewed_product': 'sum',
    'added_to_cart': 'sum',
    'checkout_started': 'sum',
    'purchase_completed': 'sum'
}).reset_index()

In [47]:
monthly_trend['conversion_rate'] = (monthly_trend['purchase_completed']/monthly_trend['visited_website']) * 100
monthly_trend

,device,campaign_type,visited_website,viewed_product,added_to_cart,checkout_started,purchase_completed,conversion_rate
0,Desktop,Discount,18162,11713,3979,2358,1214,6.684286
1,Desktop,Influencer,7035,4595,1637,983,471,6.695096
2,Desktop,New Launch,10797,7065,2524,1552,803,7.437251
3,Mobile,Discount,41959,27237,9562,5729,2893,6.894826
4,Mobile,Influencer,16780,10828,3828,2257,1155,6.883194
5,Mobile,New Launch,25267,16432,5626,3355,1645,6.510468


In [48]:
# Monthly Trend (channel + month over time)

monthly_trend = df.groupby(['month', 'channel']).agg({
    'visited_website': 'sum',
    'purchase_completed': 'sum'
}).reset_index()

In [49]:
monthly_trend['conversion_rate'] = (monthly_trend['purchase_completed']/monthly_trend['visited_website']) * 100
monthly_trend

,month,channel,visited_website,purchase_completed,conversion_rate
0,07-25,Email,2114,170,8.041627
1,07-25,Organic,6288,441,7.013359
2,07-25,Paid Ads,9222,614,6.657992
3,07-25,Social,3123,212,6.788345
4,08-25,Email,2074,172,8.293153
5,08-25,Organic,6077,413,6.796117
6,08-25,Paid Ads,9221,582,6.311680
7,08-25,Social,3117,205,6.576837
8,09-25,Email,2022,139,6.874382
9,09-25,Organic,6053,418,6.905667


In [50]:
funnel_by_channel.to_csv('pivot_01_channel.csv')
funnel_by_device.to_csv('pivot_02_device.csv')
funnel_by_user_type.to_csv('pivot_03_user_type.csv')
funnel_by_campaign.to_csv('pivot_04_campaign.csv')
funnel_by_region.to_csv('pivot_05_region.csv')
monthly_trend.to_csv('pivot_06_monthly_trend.csv')

In [51]:
df.to_csv('cleaned_data.csv', index=False)

In [52]:
df.columns

Index(['user_id', 'session_id', 'date', 'month', 'channel', 'campaign_type',
       'device', 'user_type', 'region', 'visited_website', 'viewed_product',
       'added_to_cart', 'checkout_started', 'purchase_completed',
       'discount_applied', 'order_value', 'revenue'],
      dtype='object')

# Phase 4

#### Finding 1: View → Cart is the main bottleneck

-- Evidence: 65.1% drop

-- Interpretation: Product-page engagement does not translate into cart intent.

-- Root-cause hypothesis: weak product value communication, pricing friction, or poor product-page experience.

-- Recommendation: test product-page messaging, price visibility, reviews, CTA placement, and product recommendations.

-- Validation needed: compare view→cart by channel, campaign, device, and product/category if available.

#### Finding 2: Paid Ads underperform Email

-- Evidence: Paid Ads 6.72% vs Email 7.31%.

-- Interpretation: Paid traffic is less qualified or has weaker message-to-landing-page alignment.

-- Root-cause hypothesis: targeting, creative, or landing-page mismatch.

-- Recommendation: audit paid campaigns by audience, creative, landing page, and acquisition cost.

-- Do not claim paid ads caused the gap without campaign-level evidence.

#### Finding 3: Mobile has a smaller but consistent disadvantage

-- Evidence: Mobile 6.78% vs Desktop 6.91%.

-- Stronger evidence: Email + Desktop 8.13%; Paid Ads + Mobile 6.71%.

-- Recommendation: prioritize mobile checkout and product-page QA, especially for Paid Ads.

-- Avoid calling this a major bottleneck. The gap is small.

#### Finding 4: Returning users underperform New users

-- Evidence: Returning 6.62% vs New 6.92%; Paid Ads + Returning = 6.41%.

-- Root-cause hypothesis: weak reactivation offers, irrelevant targeting, or poor returning-user experience.

-- Recommendation: test personalized offers, remarketing exclusions, and returning-user landing pages.

-- Validate with repeat-session, offer, and time-since-last-purchase data.

# RCA statement:

#### The primary funnel bottleneck is the View → Cart stage, where only 34.87% of product viewers add an item to cart.